src/v/iceberg/merge_append_action.cc

```diff
@@ -351,6 +351,160 @@ merge_append_action::upload_as_manifest(
     co_return co_await io_.upload_manifest(path, m);
 }
 
+namespace {
+
+struct promoting_visitor {
+    template<typename TType>
+    primitive_value operator()(boolean_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, boolean_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(int_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, int_type>) {
+            return v;
+        }
+        if constexpr (std::is_same_v<TType, long_type>) {
+            return long_value{v.val};
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(long_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, long_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(float_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, float_type>) {
+            return v;
+        }
+        if constexpr (std::is_same_v<TType, double_type>) {
+            return double_value{.val = v.val};
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(double_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, double_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(date_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, date_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
```

Comment:
maybe worth noting that date -> timestamp is legal in v3. don't know our 'official' story with regard to v2/v3 support, though my vague understanding is that ecosystem support for v3 is somewhat limited, currently.